RFR: Update file structure doc post major restructure (#232)
pymedphys/pymedphys: docs/developer/file-structure.rst

   pymedphys/
   |   README.rst
   |   LICENSE-AGPL-3.0-or-later
   |   LICENSE-Apache-2.0
   |   changelog.md
   |   setup.py

>    |   layers.yml

Reviewer: We can leave it there for now, but after the big split we are no longer running layer linter. So until that gets sorted should we just leave the layer description out?

:``notebooks/``: Contains a series of experimental `Jupyter notebooks`_.
                 Jupyter notebooks provide a highly convenient way to
                 experiment with code. Some of the `examples`_ are in the form
                 of Jupyter notebooks. Some PyMedPhys contributors prefer to
                 code in Jupyter notebooks to zero in on a solution before
                 attempting to add code to the PyMedPhys library. *notebooks/*
                 is a convenient place to store these experimental notebooks,
                 permitting both remote and collaborative raw development.

:``packages/``: The PyMedPhys source code library is separated into a set of
                **subpackages**, from which the main PyMedPhys package draws.

>                 Users are strictly able to install any of these subpackages,

Reviewer: ```suggestion
                Users are able to install these subpackages standalone,
```

                but almost all will instead install the main PyMedPhys
                package. The instructions given in `Installation`_ apply to the
                main PyMedPhys package. Subpackages are not really intended for

>                 user installation, but rather to help manage (the very complex)

Reviewer: ```suggestion
                user installation, but rather to provide an optional way to deploy applications with minimal dependencies when needed.
```

                     the imports included in the modules of *pymedphys* main.
                     For more details on code structure and PyMedPhys'
                     dependency handling, see :ref:`source-code`.

.. _`the PyMedPhys GitHub page`: https://github.com/pymedphys/pymedphys
.. _`examples`: ../user/examples/index.html
.. _`Installation`: ../getting-started/installation.html
.. _`Release Notes`: ../getting-started/changelog.html
.. _`Jupyter notebooks`: https://realpython.com/jupyter-notebook-introduction/
.. _`continuous integration`: https://en.wikipedia.org/wiki/Continuous_integration

.. _source-code:

The PyMedPhys Source Code
-------------------------

Almost all users will access the PyMedPhys library of source code via the
main `pymedphys` package (*pymedphys/src/pymedphys/*). However no library code
actually exists within *pymedphys/src/pymedphys/*. Instead, library code is
contained within *pymedphys/packages/* and redirected through
*pymedphys/src/pymedphys/* via a set of python imports.

Within *pymedphys/packages/*, code is organised into a set of subpackages,
such as ``pymedphys_analysis`` and ``pymedphys_dicom``. From there, each

> subpackage contains a directory named `src/<package_name>`. Within each

Reviewer: ```suggestion
subpackage contains a directory named ``src/<package_name>``. Within each
```

Do we want italics here or code blocks?

Author: I struggled to settle on a general convention, to be honest. I started using italics for paths and code blocks for module names (and code, of course), but then broke that rule further down when discussing levelling.

Maybe I'll just lock in code blocks for paths as well.

never depend on code within files of the same level, nor of higher-numbered
levels.**

Note that, in practice, "depend on" really means "import code from" using
Python's ``import`` statement.

In our example, ``g2a.py`` is in level 2, so code in ``g2a.py`` can import code
from ``g1a.py``, because ``g1a.py`` is in level 1 (a lower-numbered level). In
contrast, code in ``g2a.py`` *cannot* import code from ``g2b.py`` (which is in
the same level) or ``g3a.py`` (which is in a higher-numbered level).

*This philosophy applies for modules (categories within subpackages) as well:*
Each module has an assigned level. A module's level is flexible; it can be

> adjusted as needed. Modules are assigned levels in the file ``layers.yml``.

Reviewer: See other comment regarding layers.

View this file to see the currently assigned level of a given module. Just as
with files, modules of a given level can import from lower level modules, but
not from modules of the same or higher levels. For example, at the time of
writing, ``mudensity`` is a level 2 module, and ``gamma`` is a ``level 3``
module. This means that any file within ``gamma``, such as ``g1a.py``, is free
to import from any file within ``mudensity``, such as ``m2a.py``, regardless of
that file's level within its own module. However, no file within ``mudensity``
is allowed to import from any file in ``gamma``. Note that a module's level is
unaffected by which subpackage/s it is in.

Thankfully, we are able to programatically check for any improper file
levelling. PyMedPhys' automated test suite includes a Python package called

> ``layer-linter``, which does just that!

Reviewer: See other comment regarding layers.

For a further, in-depth explanation of the philosophy behind levelling
dependencies, see the :ref:`john-lakos` section.

> |

Reviewer: In the CSS I have now globally formatted greater section breaks by default:

https://github.com/pymedphys/pymedphys/blob/master/docs/_static/theme_overrides.css#L104-L106

I can make these section breaks even larger if you like?
